Auth pseudo login state

djbusstop opened the pull request:
Resolves: https://trello.com/c/KHG9dtGm

#### Solution
All userId state management in the authorisation context is using a new hook called `useLocalStorageUser` which is like a wrapper around useState that implements localStorage, to get the default values from localStorage, and update localStorage when the state is updated.

To represent the pseudo-login state, the state of `userId` will be the user ID acquired from local storage, authorisation, or the query string, while the isAuthenticated is `false`.

The fields available for user data is reduced in this state. Refer to **issues**.

Details:
- The `signOut` hook now uses the auth state to set the `userId` to undefined. This makes signOut work for cases where the user is authenticated, or pseudo-logged in.
- The query params are parsed in the AuthContext component, and no longer in any of the components that implemented the previous version of pseudo-login.
- `isAuthenticated` now has three states. It starts as undefined, but becomes true or false after the Amplify check. This is done to avoid repeating queries.
- `setIsAuthenticated` is now run explicitly, no longer in the `useEffect` that is listening to the state of `cognitoUser`

Side tasks:
- Moved auth hooks to their own folders to unclutter the index.js file.

#### Questions:
- Should we show the pseudologin state as "logged in" on:
    - the self scan page
    - signature list page
    - pledges components
      - It looks like the query to understand what component to show depends on the pledges data returned in the user data query, which isn't available to non logged in users. Maybe we should in these cases, show the "secret code" component? 
    - any other place `isAuthentication` is used to show the login state?
Guessing the answer is yes, since the idea is that they look logged in and then when they really want to log in, they will need to authenticate.

#### Issues:
- If we display like they are logged in, there may be cases where the logged in state requires some data that can only be accessed when the user has a token.

djbusstop commented on src/components/Forms/SelfScan/index.js:
```diff
@@ -301,15 +287,15 @@ const CountSignaturesForm = ({
   return (
     <>
-      {isAuthenticated && (
+      {userId && (
```
Shows the login state when userId is present, not just authenticated

djbusstop replied:
This works on this page, all data needed for the page is available via user api or the analytics/signatures api

djbusstop commented on src/components/Forms/Pledge/SignaturePledge.js:
```diff
@@ -136,9 +136,7 @@ export default ({ pledgeId }) => {
-                  (isAuthenticated &&
-                    userData &&
-                    !userData.newsletterConsent.value)) && (
+                  (isAuthenticated && !userData?.newsletterConsent?.value)) && (
```
Avoid errors from trying to query deep in to userData before data is availible

timurc commented:
rebasing from master should fix the build

djbusstop commented on src/context/Authentication/index.js:
```diff
@@ -13,60 +16,90 @@ import { getCurrentUser } from '../../hooks/Api/Users/Get';
+      const params = querystring.parse(window.location.search);
+      // If params, set that user as the userId and via localStorage
+      if (!userId && params.userId) {
+        setUserId(params.userId);
```
@Valioesi and I talked about this. He said to have it keep the current userId. I didn't fully get why. But I kinda see why logging out when seeing a new userId and identifying a new user makes sense.

Valioesi replied:
Ah, okay. @timurc we had a similar discussion once: what should happen if there is already an existing session when someone comes to a page like self scan with a userId in the params? Back then we decided that the user id in the params should not be regarded and the session should be used instead.

That's why I figured we would just do it the same way here. But I guess it's kind of a different situation, because we don't have real session. 

I am fine with changing it to the new user id in this case. I think it makes sense. :)

djbusstop commented on src/hooks/Authentication/index.js:
```diff
@@ -51,6 +33,11 @@ export const useSignOut = () => {
+export { useAnswerChallenge } from './AnswerChallenge';
+export { useVerification } from './Verification';
+export { useLocalStorageUser } from './LocalStorageUser';
+
```
Mostly because the `useAnswerChallenge` and `useVerification` are being used elsewhere and I wanted to break up the code without breaking the imports since those other hooks are used elsewhere.

timurc replied:
ah right. I would have changed the imports in question, but don't have a really strong option about it.

Valioesi replied:
I think the anonymized email would be a good idea for users without a username...

Valioesi replied:
Removing the user id in the local storage, if it is incorrect, would be cool actually.

Valioesi replied:
Hmh, about returning the user id after self-scanning: So far we have been trying to avoid an api which would map an email to a user id, so this would be a first. I wouldn't exactly call it a security issue, because you can't really do a whole lot with the userId (and without a token) which you wouldn't be able to do with just the email address anyway. 

I mean, our main reasoning for not mapping emails to user id was so possible attackers are not able to check if an email exists, but i guess you are able to do that anyway on mutliple occasions (e.g. when signing up through cognito, when self scanning an error is returned that a given mail does not exist, and so forth). 

In conclusion I would be open to it, if improves it UX.

djbusstop commented:
1. Fixed by not showing pledge form when user is identified on general pledge. Going to fix on signature pledge by requiring user to authenticate and then show auth state after.
2. Fixed by signing out user when user data was not able to be queried.
3. 
  - Download signature lists uses same flow. Currently ignores the pseudo log in state and uses whatever email address is supplied. Will create task.
  - Self scan works when user is identified or authenticated, and asks for email when neither. Will implement registration in the self scan in this PR https://github.com/grundeinkommensbuero/website/pull/115
4. I think this topic would be best around the topic of the self scan registration.

djbusstop commented:
Questions: 
* How should the `RequestLoginCode` component look?
  - Should it look like `AuthInfo` and use inline buttons?
  - Or use the CBA buttons?
  - Here is a photo of what I have now, which is more inline with the `AuthInfo` component. 
![angemeldetalsriver](https://user-images.githubusercontent.com/41007222/84026145-afe4b500-a984-11ea-9eef-dd559f2a8d1a.png)
* What should the text in the `RequestLoginCode` component be?

djbusstop commented:
To test this 
* Check all pledge forms have expected behaviour at different Auth levels.
  - When identified, the `SigniturePledge` will show a `RequestLoginCode` component, to get the user to authenticate. Please check that this works, including Log Out.
* Test Self Scan and Signature List pages have expected behaviour at different auth levels
* Check that old authenticated or identified user is signed out when a new userId is sent in params

djbusstop commented:
@schaerta Everything fixed or could not be recreated :)

schaerta commented:
Everything seems to be working now. It looks really cool!

Valioesi commented:
Added a fix so that users without newsletter consent receive a list after entering the sign in code. The useEffect to create the pdf is now also executed if loginCodeRequested is false.
